# Tugas Akhir Analisis Algoritma

- Muhammad Aditya Bayhaqie  : 09021282227122
- Nabila Nurhusna Yap       : 09021182227032

Assignment :
-  Benchmarking the Best Algorithm using Olist Dataset, the Listed Algorithm
  - Jump Search - Heap Sort
  - Jump Sort - Merge Sort
  - Hash Search - Selection Sort
  - Binary Search - Counting Sort

- Dataset Source [Here](https://www.kaggle.com/datasets/olistbr/brazilian-ecommerce)

## Data Collection

Let's get the Data and Library needed

In [ ]:
# Import library
import pandas as pd
import numpy as np
from zipfile import ZipFile
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from pathlib import Path
import matplotlib.pyplot as plt

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
! mkdir ~/.kaggle

In [ ]:
!cp /content/drive/MyDrive/CollabData/kaggle_API/kaggle.json ~/.kaggle/kaggle.json

In [ ]:
! chmod 600 ~/.kaggle/kaggle.json

In [ ]:
!kaggle datasets download olistbr/brazilian-ecommerce

Dataset URL: https://www.kaggle.com/datasets/fronkongames/steam-games-dataset
License(s): MIT
 90% 217M/241M [00:00<00:00, 590MB/s] 
100% 241M/241M [00:00<00:00, 591MB/s]


In [ ]:
!unzip brazilian-ecommerce.zip

Archive:  steam-games-dataset.zip
  inflating: games.csv               
  inflating: games.json              


### Data Assessment

In [ ]:
games = pd.read_csv('/content/games.csv')

In [ ]:
games.info()

<class 'pandas.core.frame.DataFrame'>
Index: 111452 entries, 20200 to 3183790
Data columns (total 39 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   AppID                       111446 non-null  object 
 1   Name                        111452 non-null  object 
 2   Release date                111452 non-null  object 
 3   Estimated owners            111452 non-null  int64  
 4   Peak CCU                    111452 non-null  int64  
 5   Required age                111452 non-null  float64
 6   Price                       111452 non-null  int64  
 7   DiscountDLC count           111452 non-null  int64  
 8   About the game              104969 non-null  object 
 9   Supported languages         111452 non-null  object 
 10  Full audio languages        111452 non-null  object 
 11  Reviews                     10624 non-null   object 
 12  Header image                111452 non-null  object 
 13  Website       

Dari data tersebut, ditarik kesimpulan bahwa:
- Terdapat 37 **Kolom**
- Terdapat 111452 **Baris**
- Beberapa Kolom memiliki **Non-null** yang sedikit (`Score rank`, `Metacritic url`, `Reviews`) dan perlu ditangani dengan beberapa metode berupa
  - Drop Kolom
  - Isi Kolom kosong

In [ ]:
# Membaca dataset

df = games
df.head(5)

AppID          Name Release date  Estimated owners  \
20200         Galactic Bowling  Oct 21, 2008    0 - 20000                 0   
655370            Train Bandit  Oct 12, 2017    0 - 20000                 0   
1732930           Jolt Project  Nov 17, 2021    0 - 20000                 0   
1355720               Henosis™  Jul 23, 2020    0 - 20000                 0   
1139950  Two Weeks in Painland   Feb 3, 2020    0 - 20000                 0   

         Peak CCU  Required age  Price  DiscountDLC count  \
20200           0         19.99      0                  0   
655370          0          0.99      0                  0   
1732930         0          4.99      0                  0   
1355720         0          5.99      0                  0   
1139950         0          0.00      0                  0   

                                            About the game  \
20200    Galactic Bowling is an exaggerated and stylize...   
655370   THE LAW!! Looks to be a showdown atop a train....   
1732930  Jolt Project: The army now has a new robotics ...   
1355720  HENOSIS™ is a mysterious 2D Platform Puzzler w...   
1139950  ABOUT THE GAME Play as a hacker who has arrang...   

                                       Supported languages  ...  \
20200                                          ['English']  ...   
655370   ['English', 'French', 'Italian', 'German', 'Sp...  ...   
1732930                 ['English', 'Portuguese - Brazil']  ...   
1355720  ['English', 'French', 'Italian', 'German', 'Sp...  ...   
1139950                     ['English', 'Spanish - Spain']  ...   

        Average playtime two weeks Median playtime forever  \
20200                            0                       0   
655370                           0                       0   
1732930                          0                       0   
1355720                          0                       0   
1139950                          0                       0   

        Median playtime two weeks             Developers  \
20200                           0  Perpetual FX Creative   
655370                          0           Rusty Moyher   
1732930                         0          Campião Games   
1355720                         0      Odd Critter Games   
1139950                         0          Unusual Games   

                    Publishers  \
20200    Perpetual FX Creative   
655370            Wild Rooster   
1732930          Campião Games   
1355720      Odd Critter Games   
1139950          Unusual Games   

                                                Categories  \
20200    Single-player,Multi-player,Steam Achievements,...   
655370   Single-player,Steam Achievements,Full controll...   
1732930                                      Single-player   
1355720              Single-player,Full controller support   
1139950                   Single-player,Steam Achievements   

                                  Genres  \
20200                Casual,Indie,Sports   
655370                      Action,Indie   
1732930  Action,Adventure,Indie,Strategy   
1355720           Adventure,Casual,Indie   
1139950                  Adventure,Indie   

                                                      Tags  \
20200                          Indie,Casual,Sports,Bowling   
655370   Indie,Action,Pixel Graphics,2D,Retro,Arcade,Sc...   
1732930                                                NaN   
1355720  2D Platformer,Atmospheric,Surreal,Mystery,Puzz...   
1139950  Indie,Adventure,Nudity,Violent,Sexual Content,...   

                                               Screenshots  \
20200    https://cdn.akamai.steamstatic.com/steam/apps/...   
655370   https://cdn.akamai.steamstatic.com/steam/apps/...   
1732930  https://cdn.akamai.steamstatic.com/steam/apps/...   
1355720  https://cdn.akamai.steamstatic.com/steam/apps/...   
1139950  https://cdn.akamai.steamstatic.com/steam/apps/...   

                                                    Movies  
20200    http://cdn.aka

Dari data tersebut, Dapat disimpulkan bahwa Terjadi pergeseran konten data dari kolom `AppID` hingga `DiscountDLCcount`, `AppID` akan di drop karena data tersebut Insignifikan dan Kolom lainnya akan direname untuk memperbaiki konten data

### Data Preparation

In [ ]:
# Rename columns
df = df.rename(columns={
    'Price': 'DiscountDLC count',
    'Required age' : 'Price',
    'Peak CCU': 'Required age',
    'Estimated owners': 'Peak CCU',
    'Release date': 'Estimated owners',
    'Name': 'Release date',
    'AppID': 'Name',
})

# Reindex the DataFrame
df = df.reset_index(drop=True)

df.head(5)

Name  Release date Estimated owners  Peak CCU  \
0       Galactic Bowling  Oct 21, 2008        0 - 20000         0   
1           Train Bandit  Oct 12, 2017        0 - 20000         0   
2           Jolt Project  Nov 17, 2021        0 - 20000         0   
3               Henosis™  Jul 23, 2020        0 - 20000         0   
4  Two Weeks in Painland   Feb 3, 2020        0 - 20000         0   

   Required age  Price  DiscountDLC count  DiscountDLC count  \
0             0  19.99                  0                  0   
1             0   0.99                  0                  0   
2             0   4.99                  0                  0   
3             0   5.99                  0                  0   
4             0   0.00                  0                  0   

                                      About the game  \
0  Galactic Bowling is an exaggerated and stylize...   
1  THE LAW!! Looks to be a showdown atop a train....   
2  Jolt Project: The army now has a new robotics ...   
3  HENOSIS™ is a mysterious 2D Platform Puzzler w...   
4  ABOUT THE GAME Play as a hacker who has arrang...   

                                 Supported languages  ...  \
0                                        ['English']  ...   
1  ['English', 'French', 'Italian', 'German', 'Sp...  ...   
2                 ['English', 'Portuguese - Brazil']  ...   
3  ['English', 'French', 'Italian', 'German', 'Sp...  ...   
4                     ['English', 'Spanish - Spain']  ...   

  Average playtime two weeks Median playtime forever  \
0                          0                       0   
1                          0                       0   
2                          0                       0   
3                          0                       0   
4                          0                       0   

  Median playtime two weeks             Developers             Publishers  \
0                         0  Perpetual FX Creative  Perpetual FX Creative   
1                         0           Rusty Moyher           Wild Rooster   
2                         0          Campião Games          Campião Games   
3                         0      Odd Critter Games      Odd Critter Games   
4                         0          Unusual Games          Unusual Games   

                                          Categories  \
0  Single-player,Multi-player,Steam Achievements,...   
1  Single-player,Steam Achievements,Full controll...   
2                                      Single-player   
3              Single-player,Full controller support   
4                   Single-player,Steam Achievements   

                            Genres  \
0              Casual,Indie,Sports   
1                     Action,Indie   
2  Action,Adventure,Indie,Strategy   
3           Adventure,Casual,Indie   
4                  Adventure,Indie   

                                                Tags  \
0                        Indie,Casual,Sports,Bowling   
1  Indie,Action,Pixel Graphics,2D,Retro,Arcade,Sc...   
2                                                NaN   
3  2D Platformer,Atmospheric,Surreal,Mystery,Puzz...   
4  Indie,Adventure,Nudity,Violent,Sexual Content,...   

                                         Screenshots  \
0  https://cdn.akamai.steamstatic.com/steam/apps/...   
1  https://cdn.akamai.steamstatic.com/steam/apps/...   
2  https://cdn.akamai.steamstatic.com/steam/apps/...   
3  https://cdn.akamai.steamstatic.com/steam/apps/...   
4  https://cdn.akamai.steamstatic.com/steam/apps/...   

                                              Movies  
0  http://cdn.akamai.steamstatic.com/steam/apps/2...  
1  http://cdn.akamai.steamstatic.com/steam/apps/2...  
2  http://cdn.akamai.steamstatic.com/steam/apps/2...  
3  http://cdn.akamai.steamstatic.com/steam/apps/2...  
4  http://cdn.akamai.steamstatic.com/steam/apps/2...  

[5 rows x 39 columns]

In [ ]:
# Drop the 8th column ('DiscountDLC count')
df = df.drop(df.columns[7], axis=1)

# Display the updated DataFrame (optional)
df.head(5)

Name  Release date Estimated owners  Peak CCU  \
0       Galactic Bowling  Oct 21, 2008        0 - 20000         0   
1           Train Bandit  Oct 12, 2017        0 - 20000         0   
2           Jolt Project  Nov 17, 2021        0 - 20000         0   
3               Henosis™  Jul 23, 2020        0 - 20000         0   
4  Two Weeks in Painland   Feb 3, 2020        0 - 20000         0   

   Required age  Price                                     About the game  \
0             0  19.99  Galactic Bowling is an exaggerated and stylize...   
1             0   0.99  THE LAW!! Looks to be a showdown atop a train....   
2             0   4.99  Jolt Project: The army now has a new robotics ...   
3             0   5.99  HENOSIS™ is a mysterious 2D Platform Puzzler w...   
4             0   0.00  ABOUT THE GAME Play as a hacker who has arrang...   

                                 Supported languages Full audio languages  \
0                                        ['English']                   []   
1  ['English', 'French', 'Italian', 'German', 'Sp...                   []   
2                 ['English', 'Portuguese - Brazil']                   []   
3  ['English', 'French', 'Italian', 'German', 'Sp...                   []   
4                     ['English', 'Spanish - Spain']                   []   

  Reviews  ... Average playtime two weeks Median playtime forever  \
0     NaN  ...                          0                       0   
1     NaN  ...                          0                       0   
2     NaN  ...                          0                       0   
3     NaN  ...                          0                       0   
4     NaN  ...                          0                       0   

  Median playtime two weeks             Developers             Publishers  \
0                         0  Perpetual FX Creative  Perpetual FX Creative   
1                         0           Rusty Moyher           Wild Rooster   
2                         0          Campião Games          Campião Games   
3                         0      Odd Critter Games      Odd Critter Games   
4                         0          Unusual Games          Unusual Games   

                                          Categories  \
0  Single-player,Multi-player,Steam Achievements,...   
1  Single-player,Steam Achievements,Full controll...   
2                                      Single-player   
3              Single-player,Full controller support   
4                   Single-player,Steam Achievements   

                            Genres  \
0              Casual,Indie,Sports   
1                     Action,Indie   
2  Action,Adventure,Indie,Strategy   
3           Adventure,Casual,Indie   
4                  Adventure,Indie   

                                                Tags  \
0                        Indie,Casual,Sports,Bowling   
1  Indie,Action,Pixel Graphics,2D,Retro,Arcade,Sc...   
2                                                NaN   
3  2D Platformer,Atmospheric,Surreal,Mystery,Puzz...   
4  Indie,Adventure,Nudity,Violent,Sexual Content,...   

                                         Screenshots  \
0  https://cdn.akamai.steamstatic.com/steam/apps/...   
1  https://cdn.akamai.steamstatic.com/steam/apps/...   
2  https://cdn.akamai.steamstatic.com/steam/apps/...   
3  https://cdn.akamai.steamstatic.com/steam/apps/...   
4  https://cdn.akamai.steamstatic.com/steam/apps/...   

                                              Movies  
0  http://cdn.akamai.steamstatic.com/steam/apps/2...  
1  http://cdn.akamai.steamstatic.com/steam/apps/2...  
2  http://cdn.akamai.steamstatic.com/steam/apps/2...  
3  http://cdn.akamai.steamstatic.com/steam/apps/2...  
4  http://cdn.akamai.steamstatic.com/steam/apps/2...  

[5 rows x 37 columns]